idaes/apps/uncertainty_propagation/uncertainties.py

```diff
@@ -0,0 +1,230 @@
+    # objective function: 
+    for cc in list(model.component_data_objects(Objective,
+                                            active=True,
+                                            descend_into=True)):
+        # save gradient of the objective with respect to only theta_names
+        if any(ele in str(cc.expr) for ele in theta_names):
+            gradient_f_theta = gradient_f[[int(line_dic[v]) -1 for v in theta_names]]
+    propagation_f = {}
+    # calculate error propagation of the objective fuction
+    if 'gradient_f_theta' in locals():
+        propagation_f['objective'] = np.dot(gradient_f_theta,np.dot(cov,np.transpose(gradient_f_theta))) + np.dot(fxxp,np.dot(cov,np.transpose(fxxp)))
+    else:
+        propagation_f['objective'] = np.dot(fxxp,np.dot(cov,np.transpose(fxxp)))
+    # constraints: 
+    # calculate error propagation dc/dp*cov_p*dc/dp + (dc/dx*dx/dp)*cov_p*(dc/dx*dx/dp)
+    num_constraints = len(list(model.component_data_objects(Constraint,
+                                                            active=True,
+                                                            descend_into=True))) 
+    if num_constraints > 0:
+        # save constraints into gradient_cc only if constraints includes theta 
+        # gradient_c[i] is a sparse matrix with column number, row number, and value
+        gradient_cc = []
+        for i in range(gradient_c.shape[0]):
+            if gradient_c[i,[0]] in line_dic.values():
+                gradient_cc.append(gradient_c[i])
+        gradient_cc = np.array(gradient_cc)
+        # save unique constraints numbers that contain theta, index starts from 1. 
+        constriant_number = list(set(gradient_cc[:,[1]].flatten().astype(int)))       
+        # convert sparse matrix to dense matrix with (value, (row number, column number))
+        I,J,V =  gradient_cc[:,[1]].flatten().astype(int)-1, gradient_cc[:,[0]].flatten().astype(int)-1,  gradient_cc[:,[2]].flatten()
+        gradient_cc = sparse.coo_matrix((V,(I,J))).todense()
+        gradient_dic = {}
+        propagation_c = {}
+        # calculate error propagation of constraints
+        for r in range(1,num_constraints+1):
+            if r in constriant_number:
+                cxxp = []
+                for i in p_list:
+                    cxxp_tmp = 0
+                    for j in x_list:
+                        cxxp_tmp = 0
+                        # we took the gradients from a sparse matrix
+                        if 'd(c'+str(r)+')/d('+j+')' in gradient_c_dic.keys():
+                            gradient_cx = gradient_c_dic['d(c'+str(r)+')/d('+j+')']
+                        else:
+                            gradient_cx = 0
+                        cxxp_tmp = cxxp_tmp + gradient_cx*dsdp_dic['d('+j+')/d('+i+')']
+                    cxxp.append(cxxp_tmp)
+                cxxp = np.array(cxxp)
+                gradient_dic['constraints '+str(r)] = gradient_cc[r-1]
```

Comment:
@Robbybp Can you offer a suggestion on how to get the name of a constraint from a Pyomo model? My recommendation is to have an option that flips between returning this with the current naming scheme OR using the actual constraint names from Pyomo.

Reply:
`r` is a row index that presumably corresponds to some constraint? And we want to access this "constraint r"? I just skimmed the code and didn't see anywhere we store or define an ordering of the constraints. We should probably create something like:
```python
constraint_list = list(model.component_data_objects(Constraint))
```
then get `constraint_list[r].name`, if we really need these constraint names. That said, getting `name` for each constraint is very expensive, so it might be better to key off something else. We could make `gradient_dic` a `ComponentMap`, then map `constraint_list[r]` to the value?

Reply:
Here is how I set the constraints name.
1. I first get [the column numbers of the theta_names in the model (=`line_dic`) and gradient dc/ds (= `gradient_c`)](https://github.com/IDAES/idaes-pse/blob/0854fab43a0e50da7860682718bdd3864fbabd9f/idaes/apps/uncertainty_propagation/uncertainties.py#L127) 
2. `gradient_c` is a sparse matrix `([colum, row, value])` from `k_aug`.
3. I take `([colum, row, value])` only include `theta_names`, and [saved it into `gradient_cc`](https://github.com/IDAES/idaes-pse/blob/0854fab43a0e50da7860682718bdd3864fbabd9f/idaes/apps/uncertainty_propagation/uncertainties.py#L163)
4. [Finally, I save unique constraint numbers that contain `theta_names`, the index starts from 1. ](https://github.com/IDAES/idaes-pse/blob/0854fab43a0e50da7860682718bdd3864fbabd9f/idaes/apps/uncertainty_propagation/uncertainties.py#L166)

I will think about it.